File: youtube/apis/channel_banners.py
"""ChannelBanners API namespace.

Wraps the following YouTube endpoint:

* POST /upload/.../channelBanners/insert — upload a channel banner image

Reference: https://developers.google.com/youtube/v3/docs/channelBanners/insert
"""

from __future__ import annotations

from youtube.apis.base import BaseAPI
from youtube.config import YOUTUBE_UPLOAD_BASE_URL
from youtube.models.channel_banners import ChannelBannerResource

_UPLOAD = YOUTUBE_UPLOAD_BASE_URL


class ChannelBannersAPI(BaseAPI):
    """Methods for uploading YouTube channel banner images."""

    async def insert(
        self,
        *,
        image_data: bytes,
        content_type: str = "image/jpeg",
    ) -> ChannelBannerResource:
        """Upload a channel banner image.

        Returns a :class:`ChannelBannerResource` whose ``url`` must be used
        in a subsequent ``channels.update`` call to set
        ``branding_settings.image.banner_external_url``.

        Parameters
        ----------
        image_data:
            Raw image bytes (JPEG or PNG, max 6 MB, 16:9 aspect ratio,
            recommended 2560 x 1440 px).
        content_type:
            MIME type — ``image/jpeg`` or ``image/png``.
        """
        payload = await self._session.upload_binary(
            f"{_UPLOAD}/channelBanners/insert",
            params={},
            data=image_data,
            content_type=content_type,
        )
        return ChannelBannerResource.model_validate(payload)

    async def insert_from_file(
        self,
        *,
        file_path: str,
        content_type: str = "image/jpeg",
    ) -> ChannelBannerResource:
        """Upload a channel banner from a local file."""
        with open(file_path, "rb") as fh:
            image_data = fh.read()
        return await self.insert(
            image_data=image_data,
            content_type=content_type,
        )

File: go/channel_banners.go
package youtube

import (
	"context"
	"errors"
	"net/url"
	"os"
)

// ChannelBannerResource is the response from InsertChannelBanner. Its URL must
// be used in a subsequent UpdateChannel call to set
// brandingSettings.image.bannerExternalUrl.
type ChannelBannerResource struct {
	Kind string `json:"kind,omitempty"`
	Etag string `json:"etag,omitempty"`
	URL  string `json:"url,omitempty"`
}

// InsertChannelBannerParams configures InsertChannelBanner.
type InsertChannelBannerParams struct {
	// ImageData is the raw image bytes (JPEG or PNG, max 6 MB, 16:9 aspect
	// ratio, recommended 2560x1440 px). Required.
	ImageData []byte
	// ContentType is the image MIME type. Defaults to "image/jpeg".
	ContentType string
}

// InsertChannelBanner uploads a channel banner image via a direct binary media
// upload. Mirrors channelBanners.insert. Requires OAuth.
func (c *Client) InsertChannelBanner(ctx context.Context, p InsertChannelBannerParams) (*ChannelBannerResource, error) {
	if err := c.requireOAuth(); err != nil {
		return nil, err
	}
	if len(p.ImageData) == 0 {
		return nil, errors.New("youtube: InsertChannelBanner requires ImageData")
	}
	contentType := p.ContentType
	if contentType == "" {
		contentType = "image/jpeg"
	}
	q := url.Values{}
	q.Set("uploadType", "media")

	out := &ChannelBannerResource{}
	if err := c.uploadMedia(ctx, "channelBanners/insert", q, p.ImageData, contentType, out); err != nil {
		return nil, err
	}
	return out, nil
}

// InsertChannelBannerFromFile uploads a channel banner from a local file.
// Mirrors channelBanners.insert_from_file. Requires OAuth.
func (c *Client) InsertChannelBannerFromFile(ctx context.Context, filePath, contentType string) (*ChannelBannerResource, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, errors.New("youtube: InsertChannelBannerFromFile: read file: " + err.Error())
	}
	return c.InsertChannelBanner(ctx, InsertChannelBannerParams{
		ImageData:   data,
		ContentType: contentType,
	})
}
